Sources/App/Core/Github.swift

```diff
@@ -67,94 +36,224 @@ enum Github {
-                        .flatMap {
-                            client.eventLoop.future(error: Error.requestFailed(.tooManyRequests, uri))
+
+    static func parseOwnerName(url: String) throws -> (owner: String, name: String) {
+        let parts = url
+            .droppingGithubComPrefix
+            .droppingGitExtension
+            .split(separator: "/")
+            .map(String.init)
+        guard parts.count == 2 else { throw Error.invalidURI(nil, url) }
+        return (owner: parts[0], name: parts[1])
+    }
+
+}
+
+
+// MARK: - GraphQL
+
+extension Github {
+
+    static let graphQLApiUri = URI(string: "https://api.github.com/graphql")
+
+    struct GraphQLQuery: Content {
+        var query: String
+    }
+
+    static func fetchResource<T: Decodable>(_ type: T.Type, client: Client, query: GraphQLQuery) -> EventLoopFuture<T> {
+        guard let token = Current.githubToken() else {
+            return client.eventLoop.future(error: Error.missingToken)
+        }
+        let headers = HTTPHeaders([("User-Agent", "SPI-Server"),
+                                   ("Authorization", "Bearer \(token)")])
+        return client.post(Self.graphQLApiUri, headers: headers) { req in
+            try req.content.encode(query)
+        }
+        .flatMap { response -> EventLoopFuture<ClientResponse> in
+            guard !isRateLimited(response) else {
+                return Current
+                    .reportError(client,
+                                 .critical,
+                                 AppError.metadataRequestFailed(nil,
+                                                                .tooManyRequests,
+                                                                Self.graphQLApiUri))
+                    .flatMap {
+                        client.eventLoop.future(error: Error.requestFailed(.tooManyRequests))
+                    }
+            }
+            return client.eventLoop.future(response)
+        }
+        .flatMapThrowing { response in
+            guard response.status == .ok else {
+                throw Error.requestFailed(response.status)
+            }
+            return try response.content.decode(T.self, using: decoder)
+        }
+    }
+
+    static func fetchMetadata(client: Client, owner: String, repository: String) -> EventLoopFuture<Metadata> {
+        struct Response: Decodable, Equatable {
+            var data: Metadata
+        }
+        return fetchResource(Response.self,
+                             client: client,
+                             query: Metadata.query(owner: owner, repository: repository))
+            .map(\.data)
+    }
+
+    static func fetchMetadata(client: Client, package: Package) -> EventLoopFuture<Metadata> {
+        do {
+            let (owner, name) = try parseOwnerName(url: package.url)
+            return fetchMetadata(client: client, owner: owner, repository: name)
```

Comment:
One thing that's a bit unfortunate is that the GraphQL query needs the parameters `owner` and `name`. Because a package only has a url we have to parse those out of the url.

Previously, we arguably weren't that much better off: we created an API url from the package url (in `static func apiUri(...)`), which also involved some parsing.

The only real difference now is that we split on the "trunk" to turn `foo/bar` into `("foo", "bar")` in addition to stripping `https://github.com` and `.git`.